### PR TITLE
Add AMQP support to TTM

### PR DESCRIPTION
**Pull request by asdil12**
See https://progress.opensuse.org/issues/27735

Example Message:

```json
'opensuse.ttm.build.failed': '{
    "project": "openSUSE:Factory",
    "build": "20180516",
    "failed_jobs": {
        "relevant": [677715, 677717, 677721, 677734, 677744, 677746, 677749, 677750, 677776, 677797, 677895, 677896, 677925, 677976, 677977, 678017],
        "ignored": [677719, 677722, 677725, 677742, 677770, 677772, 677773, 677873, 677889, 677900, 677972, 677978, 677979, 677985]
    }
}'

```

### CODE REVIEW

File: totest-manager.py [hunk 1]
```diff
@@ -895,6 +926,7 @@ def get_optparser(self):
         parser.add_option("--dry", action="store_true", help="dry run")
         parser.add_option("--debug", action="store_true", help="debug output")
         parser.add_option("--verbose", action="store_true", help="verbose")
+        parser.add_option("--config", default="~/.ttm.cfg", help="path to config file")
```
Comment (lnussel) [inline totest-manager.py]: load_first_config from xdg.BaseDirectory please

File: totest-manager.py [hunk 2]
```diff
@@ -541,6 +547,31 @@ def totest(self):
             # with a new release
             can_release = False
 
+        if self.amqp_url:
```
Comment (coolo) [inline totest-manager.py]: https://refactoring.com/catalog/extractMethod.html

**Comment by asdil12** [comment]
Updated patch.

File: totest-manager.py [hunk 3]
```diff
@@ -942,7 +979,18 @@ def _setup_totest(self, project):
                 project, ', '.join(self.totest_class))
             raise cmdln.CmdlnUserError(msg)
 
-        return self.totest_class[project](project, self.options.dry, self.options.obs_api_url, self.options.openqa_server)
+        amqp_url = None
+        configpath = os.path.expanduser(self.options.config)
```
[the remark this answers is not in the document]
Comment (asdil12) [reply totest-manager.py]: Because this is completely unrelated to osc.

**Comment by asdil12** [comment]
Updated.

**Comment by asdil12** [comment]
> why not take that from the oscrc? Avoids one more config file

@coolo ttm will now read the `ttm_amqp_url` from `oscrc`.

File: totest-manager.py [hunk 4]
```diff
@@ -575,6 +582,35 @@ def totest(self):
             self.update_totest(new_snapshot)
             self.write_version_to_dashboard("totest", new_snapshot)
 
+    def send_amqp_event(self, current_snapshot, current_result):
+        if not self.amqp_url:
+            logger.info('No ttm_amqp_url configured in oscrc - skipping amqp event emission')
```
Comment (okurz) [inline totest-manager.py]: That would trigger a lot but ok

File: totest-manager.py [hunk 5]
```diff
@@ -575,6 +582,35 @@ def totest(self):
...
+            return
+
+        logger.info('Sending AMQP message')
```
Comment (okurz) [inline totest-manager.py]: I consider this too verbose for "info" and not adding much value. How about a `logger.debug` with at least the msg_topic reflected? So one line further below

File: totest-manager.py [hunk 6]
```diff
@@ -575,6 +582,35 @@ def totest(self):
...
+        msg_topic = '%s.ttm.build.%s' % (self.project_base.lower(), self._result2str(current_result))
+        msg_body = json.dumps({
+            'build': current_snapshot,
+            'project': self.project,
+            'failed_jobs': self.failed_jobs,
+            'ignored_failed_jobs': self.ignored_failed_jobs,
+        })
```
Comment (okurz) [inline totest-manager.py]: Can you give us an example message in the PR description?

File: totest-manager.py (hunk 4 above)
Comment (coolo) [reply totest-manager.py]: that sounds indeed like a .debug

File: totest-manager.py (hunk 6 above)
Comment (asdil12) [reply totest-manager.py]: done

**Comment by okurz** [comment]
+1

**Comment by DimStar77** [comment]
> spamming IRC? :)

I hope we don't get every 5 minutes (bot run time) that the snapshot 'still failed'

**Comment by asdil12** [comment]
> I hope we don't get every 5 minutes (bot run time) that the snapshot 'still failed'

`suse_msg` has can be setup so that it will not spam similar messages

File: totest-manager.py [hunk 7]
```diff
@@ -21,6 +22,7 @@
 import signal
 import time
 import yaml
+import pika
```
Comment (DimStar77) [inline totest-manager.py]: Needs to be represented on the rpm package as a dependency (and needs to be provided on SLE12SP3, current OS of packagelist vm - currently ttm is down as the update got deployed)

**Pull request merged**